huaweicloud/services/apig/resource_huaweicloud_apig_application_quota_associate.go

```diff
@@ -0,0 +1,452 @@
+	}
+	_, err := stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return fmt.Errorf("error waiting for the binding completed: %s", err)
+	}
+	return nil
+}
+
+func appsUnbindingRefreshFunc(client *golangsdk.ServiceClient, instanceId, quota string,
+	appIds []string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		respBody, err := QueryQuotaAssociatedApplications(client, instanceId, quota)
+		if err != nil {
+			// The API returns a 404 error, which means that the instance has been deleted.
+			// In this case, there's no need to disassociate quota and application, also this action has been completed.
+			if _, ok := err.(golangsdk.ErrDefault404); ok {
+				return "instance_not_exist", "COMPLETED", nil
+			}
+			return nil, "ERROR", err
+		}
+
+		if !utils.IsSliceContainsAnyAnotherSliceElement(utils.ExpandToStringList(utils.PathSearch("[*].app_id",
+			respBody, make([]interface{}, 0)).([]interface{})), appIds, false, false) {
+			return respBody, "COMPLETED", nil
+		}
+		return "continue", "PENDING", nil
+	}
+}
+
+func resourceApplicationQuotaAssociateUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg    = meta.(*config.Config)
+		region = cfg.GetRegion(d)
+	)
+
+	client, err := cfg.NewServiceClient("apig", region)
+	if err != nil {
+		return diag.Errorf("error creating APIG client: %s", err)
+	}
+
+	var (
+		instanceId     = d.Get("instance_id").(string)
+		quotaId        = d.Get("quota_id").(string)
+		oldVal, newVal = d.GetChange("applications")
+		rmRaw          = oldVal.(*schema.Set).Difference(newVal.(*schema.Set))
+		addRaw         = newVal.(*schema.Set).Difference(oldVal.(*schema.Set))
+	)
+
+	if rmRaw.Len() > 0 {
+		if err = disassociateAppsFromQuota(ctx, client, instanceId, quotaId, parseAssociatedAppIds(rmRaw),
+			d.Timeout(schema.TimeoutUpdate)); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	if addRaw.Len() > 0 {
+		err = precheckAllAppIdsAreAvailable(ctx, client, d, parseAssociatedAppIds(addRaw), d.Timeout(schema.TimeoutUpdate))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		if err = associateAppsToQuota(ctx, client, instanceId, quotaId, parseAssociatedAppIds(addRaw),
+			d.Timeout(schema.TimeoutUpdate)); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	return resourceApplicationQuotaAssociateRead(ctx, d, meta)
+}
+
+func resourceApplicationQuotaAssociateDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg        = meta.(*config.Config)
+		region     = cfg.GetRegion(d)
+		instanceId = d.Get("instance_id").(string)
+		quotaId    = d.Get("quota_id").(string)
+	)
+
+	client, err := cfg.NewServiceClient("apig", region)
+	if err != nil {
+		return diag.Errorf("error creating APIG client: %s", err)
+	}
+
+	oldVal, _ := d.GetChange("applications")
+	if err = disassociateAppsFromQuota(ctx, client, instanceId, quotaId, parseAssociatedAppIds(oldVal.(*schema.Set)),
```

Comment:
Please precheck `oldVal.(*schema.Set).Len() > 0`

Reply:
There is unnecessary to check the length of old set. In the delete phase, if the length of old set is zero, checkdeleted logic will be triggered in the refresh step before enter delete step.